### PR TITLE
Revised Clustering Scripts

### PR DESCRIPTION
**Pull request by ddudley18**

### CODE REVIEW

File: birdclef/workflows/templates/index.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
```
Comment (acmiyaguchi) [inline birdclef/workflows/templates/index.html:1]: You may want to rename this `index.html.j2` to avoid pre-commit from breaking the templating stuff.

File: birdclef/workflows/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+)
+@click.option("--name", default=date.today())
+def main(intra, inter, checkpoint, parquet, outputdir, name):
+    model = TileNet.load_from_checkpoint(checkpoint, z_dim=512)
+    df = pd.read_parquet(parquet)
+    df["species"] = df.source_name.apply(lambda x: x.split("/")[1])
+    df[["species"]].groupby("species").size().sort_values(ascending=False)
+
+    data = []
+    root = Path("../../data/raw/birdclef-2022")
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:56]: Make this an option

File: birdclef/workflows/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+@click.option(
+    "--outputdir",
+    type=click.Path(exists=True, dir_okay=True),
+    default=Path(
+        "../../data/intermediate/generated_resources"
+    ),
+)
+@click.option("--name", default=date.today())
+def main(intra, inter, checkpoint, parquet, outputdir, name):
+    model = TileNet.load_from_checkpoint(checkpoint, z_dim=512)
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:50]: Make the dim an option too

File: birdclef/workflows/evaluation.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    data = []
+    root = Path("../../data/raw/birdclef-2022")
+    for row in df[df.species == intra].sample(50).itertuples():
+        y = motif.load_audio(root / row.source_name, int(row.motif_0), 5)
+        data.append(y)
+
+    z = data[0]
+    plt.plot(z)
+    os.makedirs(outputdir, exist_ok=True)
+    plt.savefig(outputdir + '/' + name + '/' + intra + '1.png', bbox_inches='tight')
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:64]:
```suggestion
    plt.savefig(f"{outputdir}/{name}/{intra}1.png", bbox_inches='tight')
```
Format strings are nicer that using `+` to concatenate strings (imo)

File: birdclef/workflows/evaluation.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    plt.clf()
+
+    kmeans = KMeans(n_clusters=3).fit(emb)
+    kmeans.labels_
+    plt.scatter(g[:, 0], g[:, 1], c=kmeans.labels_)
+    plt.savefig(outputdir + '/' + name + '/' + intra + '3.png', bbox_inches='tight')
+    intrafigures.append(outputdir + '/' + name + '/' + intra + '3.png')
+    plt.clf()
+
+    # Interclass Clustering and Comparison
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:84]: You might want to add some print statements to give a sense of progress.

File: birdclef/workflows/evaluation.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    le = LabelEncoder().fit(labels)
+
+    g = PCA(n_components=2).fit_transform(emb)
+    plt.scatter(g[:, 0], g[:, 1], c=le.transform(labels))
+    store = outputdir + '/' + name + '/' + inter + '1.png'
+    plt.savefig(store, bbox_inches='tight')
+    plt.clf()
+
+    # Create html file with Jinja2 template
+    file_loader = FileSystemLoader('templates')
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:109]: Does the script actually find the correct templates directory?

File: birdclef/workflows/evaluation.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    type=click.Path(exists=True, dir_okay=False),
+    default=Path(
+        "../../data/processed/2022-04-03-motif-consolidated.parquet"
+    ),
+)
+@click.option(
+    "--outputdir",
+    type=click.Path(exists=True, dir_okay=True),
+    default=Path(
+        "../../data/intermediate/generated_resources"
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:45]: You might want to make this relative to the file e.g. make a `ROOT` variable in the toplevel namespace.

```
ROOT = Path(__file__).parent.parent
```

And then you can have this be: `f"{ROOT}/data/intermediate/generated_resources"`

**Comment by acmiyaguchi** [comment]
Make sure that this passes the pre-commit checks, you can run `pre-commit run -a` to ensure that the files are formatted correctly. Make sure to rename the index.html with a different ending so it doesn't get mangled by prettier.

File: birdclef/workflows/evaluation.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+
+@evaluation.command()
+@click.option('--intra', default="brnowl", help='Singular species to perform intraclass inspection')
+@click.option('--inter', default="brnowl,skylar,norcar", help='Array of 3 species to perform interclass comparison')
+@click.option(
+    "--checkpoint",
+    type=click.Path(exists=True, dir_okay=False),
+    default=Path(
+        "../../data/processed/model/2022-04-12-v4/embedding.ckpt"
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:31]: 2022-04-12-v4 is old and won't work with current versions of the code, you'll need to use a more recent model.

File: birdclef/workflows/evaluation.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    for row in df[df.species.isin(inter.split(","))].sample(300).itertuples():
+        y = motif.load_audio(root / row.source_name, int(row.motif_0), 5)
+        data.append(y)
+        labels.append(row.species)
+
+    for i in range(3):
+        z = data[i]
+        plt.plot(z)
+        plt.clf()
+        ipd.display(ipd.Audio(z, rate=32000))
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:108]: remove this command, since we dont want to pull in jupyter dependencies into the main module

File: birdclef/workflows/evaluation.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
+import os
+import webbrowser
+from datetime import date
+from pathlib import Path
+
+import click
+import IPython.display as ipd
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:7]: remove this dependency

File: birdclef/workflows/evaluation.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+    template = env.get_template("index.html.j2")
+    output = template.render(
+        intraspecies=intra,
+        interspecies=inter,
+        intrafigures=intrafigures,
+        interfigures=store,
+    )
+
+    with open(f"{outputdir}/{name}/index.html.j2", "w") as fp:
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:132]: The output of the rendered template can be named `index.html`, j2 stands for Jinja2 (the template name)

File: birdclef/workflows/evaluation.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+    g = PCA(n_components=2).fit_transform(emb)
+    plt.scatter(g[:, 0], g[:, 1], c=le.transform(labels))
+    store = f"{outputdir}/{name}/{inter}1.png"
+    plt.savefig(store, bbox_inches="tight")
+    plt.clf()
+    print("Generated Interclass Figures")
+
+    # Create html file with Jinja2 template
+    file_loader = FileSystemLoader("../../birdclef/workflows/templates")
```
Comment (acmiyaguchi) [inline birdclef/workflows/evaluation.py:121]: This won't work as expected because we're in a python package, use importlib.resources:

https://stackoverflow.com/questions/6028000/how-to-read-a-static-file-from-inside-a-python-package

**Review by acmiyaguchi — APPROVED** [review]
This looks good to me, thanks for all the back and forth on this.
